### ISSUE TITLE
Crash when using a combination of a switch statement, optional chaining, an implicit any array, and ts strict settings.

### ISSUE THREAD

**Issue by VincentGuinaudeau** [microsoft/TypeScript#44659]
# Bug Report

I discovered this crash today, when trying to build my JS project with typescript (I want to start introducing some TS files).

Once I refined the reproduction scenario, I was able to make any version of typescript crash, on any environment (playground, CLI, LSP).

### 🔎 Search Terms

crash, strict, optional chaining, implicit any

### 🕗 Version & Regression Information

- This is a crash
- This occurs on versions 3.7.2 and onward (tried with @next)
- 3.6.5 shows diagnotics, but doesn't handle optional chaining

### ⏯ Playground Link

No diagnostic is shown, despite the array missing a type annotation.

The crash can be fixed by :
 * adding a type annotation to the array (even `any []`)
 * removing the optional chaining
 * removing the access to `foo[0]`
 * turning the `switch` into an `if`
 * turning off `strictNullChecks`
 * turning off `noImplicitAny`

https://www.typescriptlang.org/play?&strict=true#code/DYUwLgBAZg9jEF4IG0C6BYAUFgzgdwEswBjACwgApYYB+AOlADsBzMUgSggG8sBIYgIY4QEAIwAuPr2rIADBkwBfIA

### 💻 Code

I know full well how bad this code is, it has already been fixed.

Pinpointing the issue was hard though because typescript doesn't provide any human readable trace as to what code it was parsing when the error occurred.

`tsconfig.json`
```json
{
	"compilerOptions": {
		"outDir": "./dist",

		"strictNullChecks": true,
		"noImplicitAny": true
	},
	"include": ["index.ts"]
}
```

`index.ts`
```typescript
let foo = []

switch (foo?.length) {
	case 1:
		foo[0]
}
```

```
# with typescript 4.3.4
$ tsc
$HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:90672
                throw e;
                ^

TypeError: Cannot read property 'length' of undefined
    at isFunctionObjectType ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54183:47)
    at getTypeFacts ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54227:21)
    at $HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54260:60
    at filterType ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54485:43)
    at getTypeWithFacts ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54260:20)
    at narrowTypeBySwitchOptionalChainContainment ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:55419:44)
    at getTypeAtSwitchClause ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:55076:36)
    at getTypeAtFlowNode ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54893:32)
    at getFlowTypeOfReference ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:54840:51)
    at checkIdentifier ($HOME/.npm_global/lib/node_modules/typescript/lib/tsc.js:55902:28)
```
